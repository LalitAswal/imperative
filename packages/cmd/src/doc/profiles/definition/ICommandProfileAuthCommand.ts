/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandExampleDefinition, ICommandOptionDefinition } from "../../../..";

export interface ICommandProfileAuthCommand {
    /**
     * Path to command handler
     */
    handler: string;

    /**
     * Command description for help text
     */
    description?: string;

    /**
     * Command examples for help text
     */
    examples?: ICommandExampleDefinition[];

    /**
     * Additional command options
     */
    options?: ICommandOptionDefinition[];
}
