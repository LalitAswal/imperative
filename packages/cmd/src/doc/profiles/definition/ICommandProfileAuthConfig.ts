/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { TOKEN_TYPE_CHOICES } from "../../../../../rest/src/session/SessConstants";
import { IPartialCommandDefinition } from "../../IPartialCommandDefinition";

export interface ICommandProfileAuthConfig {
    /**
     * Name of the authentication service
     */
    serviceName: string;

    /**
     * Path to the handler for the authentication service.
     * The handler should inherit from Imperative BaseAuthHandler.
     */
    handler: string;

    /**
     * Type of token used by the authentication service
     */
    tokenType?: TOKEN_TYPE_CHOICES;

    /**
     * Command properties for `auth login <serviceName>`
     */
    login?: IPartialCommandDefinition;

    /**
     * Command properties for `auth logout <serviceName>`
     */
    logout?: IPartialCommandDefinition;
}
