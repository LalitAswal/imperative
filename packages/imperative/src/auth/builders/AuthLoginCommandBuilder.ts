/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { AuthCommandBuilder } from "./AuthCommandBuilder";
import { ICommandDefinition } from "../../../../cmd";
import { loginAuthCommandDesc } from "../../../../messages";
import { Constants } from "../../../../constants";
import { TextUtils } from "../../../../utilities";
import { Logger } from "../../../../logger/index";
import { ProfilesConstants } from "../../../../profiles";
import { ICommandProfileAuthConfig } from "../../../../cmd/src/doc/profiles/definition/ICommandProfileAuthConfig";

/**
 * Used to build auth login command definitions.
 * Used automatically if you allow the "auth" command group to be generated
 */
export class AuthLoginCommandBuilder extends AuthCommandBuilder {
    /**
     * Gets the "action" that this command builder is building.
     * @return {string}: The "login" action string
     */
    public getAction(): string {
        return Constants.LOGIN_ACTION;
    }

    /**
     * Build the full command - includes action group and object command.
     * @return {ICommandDefinition}: The command definition.
     */
    public buildFull(): ICommandDefinition {
        return this.buildAuthSegmentFromConfig();
    }

    /**
     * Builds only the "auth" segment from the auth config.
     * @return {ICommandDefinition}
     */
    protected buildAuthSegmentFromConfig(): ICommandDefinition {
        const authType: string = this.mConfig.serviceName;
        const authCommand: ICommandDefinition = {
            name: authType,
            summary: TextUtils.formatMessage(loginAuthCommandDesc.message,
                {type: authType}),
            description: this.mConfig.login.description,
            type: "command",
            handler: this.mConfig.login.handler,
            options: this.mConfig.login.options,
            examples: this.mConfig.login.examples,
            profile: {
                optional: [this.mProfileType]
            },
            customize: {}
        };
        // authCommand.customize[ProfilesConstants.PROFILES_COMMAND_TYPE_KEY] = this.mProfileType;

        if (authCommand.description == null) {
            authCommand.description = authCommand.summary;
        }
        return authCommand;
    }
}
