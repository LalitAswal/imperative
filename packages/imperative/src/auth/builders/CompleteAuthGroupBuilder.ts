/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { ICommandDefinition } from "../../../../cmd";
import {
    loginAuthsCommandDesc, loginAuthsCommandSummary,
    logoutAuthsCommandDesc, logoutAuthsCommandSummary
} from "../../../../messages";
import { Constants } from "../../../../constants";
import { AuthLoginCommandBuilder } from "./AuthLoginCommandBuilder";
import { AuthLogoutCommandBuilder } from "./AuthLogoutCommandBuilder";
import { Logger } from "../../../../logger/index";
import { ICommandProfileAuthConfig } from "../../../../cmd/src/doc/profiles/definition/ICommandProfileAuthConfig";

/**
 * Generate a complete group of commands for logging in and out of services
 * based on provided auth definitions.
 */
export class CompleteAuthGroupBuilder {
    /**
     * Get the complete auth group of commands
     * @param {[key: string]: ICommandProfileAuthConfig} authConfigs - mapping of profile types to auth configs
     * @param {Logger} logger - logger to use in the builder classes
     * @returns {ICommandDefinition} - the complete profile group of commands
     */
    public static getAuthGroup(authConfigs: {[key: string]: ICommandProfileAuthConfig[]}, logger: Logger): ICommandDefinition {
        const authGroup: ICommandDefinition = {
            name: Constants.AUTH_GROUP,
            description: "Manage tokens for authentication services",
            type: "group",
            children: []
        };

        const loginGroup: ICommandDefinition = {
            name: Constants.LOGIN_ACTION,
            description: loginAuthsCommandDesc.message,
            summary: loginAuthsCommandSummary.message,
            aliases: ["li"],
            type: "group",
            children: [],
        };

        const logoutGroup: ICommandDefinition = {
            name: Constants.LOGOUT_ACTION,
            description: logoutAuthsCommandDesc.message,
            summary: logoutAuthsCommandSummary.message,
            aliases: ["lo"],
            type: "group",
            children: [],
        };

        const cmdGroups: ICommandDefinition[] = [];
        for (const profileType of Object.keys(authConfigs)) {
            for (const authConfig of authConfigs[profileType]) {
                const loginCommandAction = new AuthLoginCommandBuilder(profileType, logger, authConfig);
                const logoutCommandAction = new AuthLogoutCommandBuilder(profileType, logger, authConfig);
                loginGroup.children.push(loginCommandAction.build());
                logoutGroup.children.push(logoutCommandAction.build());
            }
        }
        authGroup.children.push(loginGroup, logoutGroup);
        return authGroup;
    }
}
