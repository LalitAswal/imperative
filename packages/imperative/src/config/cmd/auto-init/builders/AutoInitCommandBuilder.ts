/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { AbstractCommandBuilder } from "../../../../../../cmd/src/builders/AbstractCommandBuilder";
import { ICommandDefinition } from "../../../../../../cmd";
import { Logger } from "../../../../../../logger";
import { ICommandProfileAutoInitConfig } from "../../../../../../cmd/src/doc/profiles/definition/ICommandProfileAutoInitConfig";
import { ImperativeError } from "../../../../../../error";
import { TextUtils } from "../../../../../../utilities";
import { autoInitCommandDesc } from "../../../../../../messages";
import { Constants } from "../../../../../../constants";

/**
 * Class for generating auth-related commands
 */
export class AutoInitCommandBuilder implements AbstractCommandBuilder {

    /**
     * Auth config for the command.
     */
    protected mConfig: ICommandProfileAutoInitConfig;

    /**
     * Construct the builder based on the auth config.
     * @param mProfileType - the profile name of the profile type e.g. banana
     * @param {Logger} mLogger - logger instance to use for the builder class
     * @param {IImperativeAuthConfig} mAuthConfig - the config for the auth type
     */
    constructor(protected mLogger: Logger,
                protected mAutoInitConfig: ICommandProfileAutoInitConfig,
                protected mProfileType?: string) {

        this.mConfig = mAutoInitConfig;
        if (this.mConfig == null) {
            throw new ImperativeError({msg: `Auto-init Builder Error: No auto-init config was supplied.`});
        }
    }

    /**
     * Build the full command - includes action group and object command.
     * @return {ICommandDefinition}: The command definition.
     */
    public buildFull(): ICommandDefinition {
        return this.buildAutoInitSegmentFromConfig();
    }

    /**
     * Gets the "action" that this command builder is building.
     * @return {string}: The auth action string
     */
    public getAction(): string {
        return Constants.AUTO_INIT_ACTION;
    }

    /**
     * Only constructs the "group" command segment for the document. Use this if the command definition
     * document already includes an auth verb.
     * @return {ICommandDefinition}
     */
    public build(): ICommandDefinition {
        return this.buildAutoInitSegmentFromConfig();
    }
    /**
     * Builds only the "auto-init" segment from the auto-init config.
     * @return {ICommandDefinition}
     */
    protected buildAutoInitSegmentFromConfig(): ICommandDefinition {
        const authCommand: ICommandDefinition = {
            name: "auto-init",
            type: "command",
            summary: this.mConfig.autoInit?.summary,
            description: this.mConfig.autoInit?.description,
            handler: this.mConfig.handler,
            options: [
                ...(this.mConfig.autoInit?.options || [])
            ],
            examples: this.mConfig.autoInit?.examples,
            profile: {
                optional: [this.mProfileType]
            },
            customize: {}
        };

        if (authCommand.summary == null) {
            authCommand.summary = TextUtils.formatMessage(autoInitCommandDesc.message, {source: this.mConfig.provider});
        }
        if (authCommand.description == null) {
            authCommand.description = authCommand.summary;
        }
        return authCommand;
    }
}
