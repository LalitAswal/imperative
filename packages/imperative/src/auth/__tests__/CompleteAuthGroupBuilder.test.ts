import { CompleteAuthGroupBuilder } from "../builders/CompleteAuthGroupBuilder";
import { Logger } from "../../../../logger";
import { ICommandProfileAuthConfig } from "../../../../cmd/src/doc/profiles/definition/ICommandProfileAuthConfig";
import { ICommandDefinition } from "../../../../cmd";
import { fakeAuthConfig } from "./__data__/SampleAuthConfig";

const authConfigs: {[key: string]: ICommandProfileAuthConfig[]} = {
    base: [ fakeAuthConfig ]
};

describe("CompleteAuthGroupBuilder", () => {
    it("should create complete auth group given an auth config object", () => {
        const cmdDef: ICommandDefinition = CompleteAuthGroupBuilder.getAuthGroup(authConfigs, Logger.getImperativeLogger());
        expect(cmdDef).toMatchSnapshot();
    });
});
