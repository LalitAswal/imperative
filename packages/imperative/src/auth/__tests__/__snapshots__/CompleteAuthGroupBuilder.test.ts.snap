// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompleteAuthGroupBuilder should create complete auth group given an auth config object 1`] = `
Object {
  "children": Array [
    Object {
      "aliases": Array [
        "li",
      ],
      "children": Array [
        Object {
          "customize": Object {},
          "description": "Fake login command",
          "examples": Array [
            Object {
              "description": "Fake login example",
              "options": "",
            },
          ],
          "handler": "fakeHandler",
          "name": "fakeService",
          "options": Array [
            Object {
              "aliases": Array [
                "st",
              ],
              "description": "Show the token when login is successful. If specified, does not save the token to a profile.",
              "name": "show-token",
              "type": "boolean",
            },
            Object {
              "aliases": Array [
                "cp",
              ],
              "description": "Create a default profile with the name 'default'. Will fail if a profile with the name 'default' already exists.",
              "name": "create-profile",
              "type": "boolean",
            },
            Object {
              "description": "Fake command option",
              "name": "fake",
              "type": "string",
            },
          ],
          "profile": Object {
            "optional": Array [
              "base",
            ],
          },
          "summary": "Login to a fakeService instance",
          "type": "command",
        },
      ],
      "description": "Login to an authentication server.",
      "name": "login",
      "summary": "Login to an authentication server",
      "type": "group",
    },
    Object {
      "aliases": Array [
        "lo",
      ],
      "children": Array [
        Object {
          "customize": Object {},
          "description": "Fake logout command",
          "examples": Array [
            Object {
              "description": "Fake logout example",
              "options": "",
            },
          ],
          "handler": "fakeHandler",
          "name": "fakeService",
          "options": Array [
            Object {
              "description": "Fake command option",
              "name": "fake",
              "type": "string",
            },
          ],
          "profile": Object {
            "optional": Array [
              "base",
            ],
          },
          "summary": "Logout of a fakeService instance",
          "type": "command",
        },
      ],
      "description": "Logout of an authentication server.",
      "name": "logout",
      "summary": "Logout of an authentication server",
      "type": "group",
    },
  ],
  "description": "Manage tokens for authentication services",
  "name": "auth",
  "type": "group",
}
`;
